mne/stats/regression.py

```diff
@@ -134,3 +142,194 @@ def _fit_lm(data, design_matrix, names):
+    # time lags. Thus, each array is mostly sparse, with one diagonal of 1s
+    # per event (for binary predictors).
+
+    # This should probably be improved (including making it more robust to
+    # high frequency data) by operating on sparse matrices. As-is, high
+    # sampling rates plus long windows blow up the system due to the inversion
+    # of massive matrices.
+    # Furthermore, assigning to a preallocated array would be faster.
+
+    cond_length = dict()
+    pred_arrays = list()
+    for cond in conds:
+
+        # create the first row and column to be later used by toeplitz to build
+        # the full predictor matrix
+        tmin_, tmax_ = tmin[cond], tmax[cond]
+        n_lags = int(tmax_ - tmin_)
+        samples, lags = np.zeros(len(times)), np.zeros(n_lags)
+
+        if cond in event_id.keys():  # for binary predictors
+            ids = ([event_id[cond]] if isinstance(event_id[cond], int)
+                   else event_id[cond])
+            samples[np.asarray([t for t, _, e in events
+                                if e in ids]) + int(tmin_)] = 1  # np.in1d
+
+        else:  # for predictors from covariates, e.g. continuous ones
+            if len(covariates[cond]) != len(events):
+                error = """Condition {} from ```covariates``` is
+                        not the same length as ```events```""".format(cond)
+                raise ValueError(error)
+            for tx, v in zip(events[:, 0], covariates[cond]):
+                samples[tx + int(tmin_)] = np.float(v)
+
+        cond_length[cond] = len(np.nonzero(samples))
+
+        # this is the magical part (thanks to Marijn van Vliet):
+        # use toeplitz to construct series of diagonals
+        pred_arrays.append(linalg.toeplitz(samples, lags))
+
+    big_arr = np.hstack(pred_arrays).T
+    # find only those positions where at least one predictor isn't 0
+    has_val = np.asarray([len(np.nonzero(line)[0]) > 0
+                          for line in big_arr.T])
+
+    # additionally, reject positions based on extreme steps in the data
+    if reject is not None:
+        from ..utils import _reject_data_segments
+        if not isinstance(reject, dict):
+            reject = dict(grad=4000e-12,  # T / m (gradiometers)
+                          mag=4e-11,  # T (magnetometers)
+                          eeg=40e-5,  # uV (EEG channels)
+                          eog=250e-5  # uV (EOG channels)
+                          )
+
+        _, inds = _reject_data_segments(data, reject, None, None,
+                                        raw.info, tstep)
+        for t0, t1 in inds:
+            has_val[t0:t1] = False
+
+    X = big_arr[:, has_val].T
+#    X = np.vstack((X, np.ones(X.shape[1]))).T  # currently no intercept
+
+    Y = data[:, has_val]
+
+    # solve linear system
+    # note: inv is slightly (~10%) faster, but pinv seemingly more stable
+    coefs = fast_dot(linalg.pinv(fast_dot(X.T, X)), fast_dot(X.T, Y.T)).T
```

Comment:
fine to tackle sparse matrices later but the cosmetics should be done before merge.